lib/cmsis_rtos_v1/cmsis_thread.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <atomic.h>
+#include <init.h>
+#include <device.h>
+#include "cmsis_thread.h"
+
+struct work_item {
+	struct cmsis_thread *thrd;
+	struct k_delayed_work work;
+};
+static struct work_item worker;
+
+static void work_handler(struct k_work *work)
+{
+	struct work_item *ti = CONTAINER_OF(work, struct work_item, work);
+
+	/*Free the mempool here */
+	k_mem_pool_free(&ti->thrd->block);
+}
+
+static int init_cmsis_rtos_v1_post_kernel(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	k_work_init(&worker.work.work, work_handler);
+	return 0;
+}
+
+SYS_INIT(init_cmsis_rtos_v1_post_kernel,
+		POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+static s32_t zephyr_to_cmsis_priority(u32_t z_prio)
+{
+	return(osPriorityRealtime - z_prio);
+}
+
+static u32_t cmsis_to_zephyr_priority(s32_t c_prio)
+{
+	return(osPriorityRealtime - c_prio);
+}
+
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	osThreadId thread_id;
+	struct cmsis_thread *cm_thread;
+	void * (*fun_ptr)(void *) = arg3;
+
+	fun_ptr(arg1);
+
+	thread_id = osThreadGetId();
+
+	/* The below operation gets a pointer to the parent structure
+	 * of the thread which also has k_mem_block as one of its elements.
+	 * This needs to be used to free the mem pool.
+	 */
+	cm_thread = CONTAINER_OF(thread_id, struct cmsis_thread, thread);
+
+	worker.thrd = cm_thread;
+	k_work_submit(&worker.work.work);
```

Comment:
I see what you're doing here, but this trick isn't as safe as it looks.  You want to guarantee that the thread is truly finished executing before you do the free of its thread struct.  But a work queue doesn't guarantee that.  If someone has played with priorities (or if you're on an SMP system, though AFAIK there are no plans for CMSIS on Cortex A) you could theoretically find the work queue executing before the thread is finished.

Getting this right is going to be tricky.  Some combination of locking interrupts, giving a semaphore and then calling k_thread_abort(k_thread_current()) is probably going to be needed.

Reply:
Yes. We too have noticed this. In fact, I have sent a mail on this (you are included). IMO, this can be addressed in 2 ways.
a.	Decrease the priority of the work queue thread (default is -1) to that of the Idle thread. Then, when a thread is aborted, it’s state is set accordingly. The work queue thread can look at this state and then free the corresponding memory occupied by the thread.
b.	Have a new thread of Idle priority which will do the above job. Something like a clean-up thread.